runtime/src/iree/hal/drivers/hip/event_semaphore.c

```diff
@@ -99,484 +132,749 @@ static void iree_hal_hip_semaphore_destroy(
   IREE_TRACE_ZONE_END(z0);
   return status;
 }
 
-static void iree_hal_hip_semaphore_fail(iree_hal_semaphore_t* base_semaphore,
-                                        iree_status_t status) {
+static iree_status_t iree_hal_hip_event_semaphore_advance(
+    iree_hal_semaphore_t* base_semaphore) {
   iree_hal_hip_semaphore_t* semaphore =
       iree_hal_hip_semaphore_cast(base_semaphore);
   IREE_TRACE_ZONE_BEGIN(z0);
 
-  const iree_status_code_t status_code = iree_status_code(status);
-
-  iree_slim_mutex_lock(&semaphore->mutex);
+  iree_hal_hip_semaphore_work_item_t* work_item = NULL;
+  iree_hal_hip_semaphore_work_item_t* last_work_item = NULL;
 
-  // Try to set our local status - we only preserve the first failure so only
-  // do this if we are going from a valid semaphore to a failed one.
-  if (!iree_status_is_ok(semaphore->failure_status)) {
-    // Previous sta-tus was not OK; drop our new status.
-    IREE_IGNORE_ERROR(status);
+  // Take out all of the values from the queue that are less than the
+  // current visible value, and make sure we advance any work needed
+  // on them.
+  do {
+    iree_slim_mutex_lock(&semaphore->mutex);
+    iree_hal_hip_util_tree_node_t* node =
+        iree_hal_hip_util_tree_first(&semaphore->event_queue.tree);
+    if (node == NULL) {
+      iree_slim_mutex_unlock(&semaphore->mutex);
+      break;
+    }
+    if (iree_hal_hip_util_tree_node_get_key(node) >
+        semaphore->current_visible_value) {
+      iree_slim_mutex_unlock(&semaphore->mutex);
+      break;
+    }
+    iree_hal_hip_semaphore_queue_item_t copy =
+        *(iree_hal_hip_semaphore_queue_item_t*)
+            iree_hal_hip_util_tree_node_get_value(node);
+    iree_hal_hip_util_tree_erase(&semaphore->event_queue.tree, node);
     iree_slim_mutex_unlock(&semaphore->mutex);
-    IREE_TRACE_ZONE_END(z0);
-    return;
-  }
+    if (copy.event) {
+      iree_hal_hip_event_release(copy.event);
+    }
+    if (copy.cpu_event) {
+      iree_event_set(&copy.cpu_event->event);
+      iree_hal_resource_release(&copy.cpu_event->resource);
+    }
 
-  // Signal to our failure sentinel value.
-  semaphore->current_value = IREE_HAL_SEMAPHORE_FAILURE_VALUE;
-  semaphore->failure_status = status;
+    iree_hal_hip_semaphore_work_item_t* next_work_item = copy.work_item;
+    while (next_work_item) {
+      if (!work_item) {
+        work_item = next_work_item;
+      }
+      if (last_work_item && !last_work_item->next) {
+        last_work_item->next = next_work_item;
+      }
+      last_work_item = next_work_item;
+      next_work_item = next_work_item->next;
+    }
+  } while (true);
 
+  iree_slim_mutex_lock(&semaphore->mutex);
+  semaphore->max_value_to_be_signaled = iree_max(
+      semaphore->max_value_to_be_signaled, semaphore->current_visible_value);
+  iree_status_t status = iree_status_clone(semaphore->failure_status);
   iree_slim_mutex_unlock(&semaphore->mutex);
+  // Now that we have accumulated all of the work items, and we have
```

Comment:
splitting code under a lock from code that is not is a good use of vertical whitespace - not having it here makes it harder to spot that the unlock happens
```suggestion

  // Now that we have accumulated all of the work items, and we have
```